### PR TITLE
Main break argparse

### PR DESCRIPTION
**Pull request by Simontwice**
# Description
reorganizes argparse, and makes crucial arguments non-default

# Checklist
Answer each question with "yes", and provide an explanation if the answer isn't what is expected.

1. Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - Answer: yes
2. Are all variables set with logical defaults that are backward compatible?
   - Answer:yes
3. Can you provide a link to a Neptune experiment verifying the functionality of this PR?
   - Answer: no point
4. Have you successfully executed the linter and tests?
   - Answer:yes
5. Are all functions concise and don't require splitting?
   - Answer:yes
6. Is there documentation for arguments and complex logic?
   - Answer:yes
7. Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - Answer:yes
8. Are all functions placed in appropriate files?
   - Answer:yes
9. Is the PR name meaningful and descriptive?
   - Answer:yes
10. Is PR description provided, or unnecessary?
    - Answer:yes
  
Other questions:

11. Have you not made changes to the requirements or anything related to the Singularity image? If you have, did you generate and upload a new image to the clusters?
    - Answer:
12. Is the change significant enough to notify all individuals working with this repository?
    - Answer:
13. Is there a need for a second reviewer?
    - Answer:

### CODE REVIEW

File: research/conditional/utils/argparse.py [hunk 1]
```diff
@@ -4,74 +4,116 @@
...
+    parser.add_argument("--data_distributed", action="store_true")
+    parser.add_argument(
+        "--model_parallelism_fragmentation",
+        type=str,
+        default=None,
+        help="comma-separated list of integers, that signify the numbers of model blocks that are first on the new device, e.g. 2,4 means that blocks 0,1 will be on GPU 0, blocks 2,3 will be on GPU 1, and the rest will be on GPU 2",
+    )
+    parser.add_argument("--detect_anomaly", action="store_true")
+
+    # CORE data hyperparameters, almost always specified in baseline configs
```
Comment (sebastianjaszczur) [inline research/conditional/utils/argparse.py:68]: i'd put all of those above model / training hyperparams

File: research/conditional/utils/argparse.py [hunk 2]
```diff
@@ -4,74 +4,116 @@
...
     parser.add_argument("--save_weights_interval", type=int, default=1000)
     parser.add_argument("--load_weights_path", type=str, default=None)
-    parser.add_argument("--grad_clip", type=float, default=None)
-    parser.add_argument("--weight_decay", type=float, default=0.0)
-    parser.add_argument("--adam_beta1", type=float, default=0.9)
-    parser.add_argument("--adam_beta2", type=float, default=0.999)
-    parser.add_argument("--no_ff", action="store_true")
-    parser.add_argument("--loss_checkpoint_chungs", type=int, default=0)
-    parser.add_argument("--gradient_accumulation_steps", type=int, default=1)
-    parser.add_argument("--auto_find_grad_accumulation", action="store_true")
```
Comment (sebastianjaszczur) [inline research/conditional/utils/argparse.py:49]: this parameter got removed completely?

**Review by sebastianjaszczur — COMMENTED** [review]
see minor comments. Other than that, I'd feel more comfortable is also somebody else looked at that PR before merging and approved it. So i just send comments, but it LGTM. Thanks!

File: research/conditional/utils/argparse.py (hunk 2 above)
Comment (Simontwice) [reply research/conditional/utils/argparse.py:49]: correct, though it shouldve been done when I removed the logic for that as well. An oversight on my side.

File: research/conditional/train/configs/test/test_baseline.yaml [hunk 3]
```diff
@@ -30,4 +30,7 @@ params:
   lr_decay: 0.8
   lr_warmup_steps: 0
   lr_decay_interval: 10000
+  n_att_heads: 4
+  learning_rate: 0.0001
```
Comment (crewtool) [inline research/conditional/train/configs/test/test_baseline.yaml:34]: default jest inny, więc to się łączy z taskiej ustalania test_baseline'u.
Zmiana poza scope'm

File: research/conditional/utils/argparse.py [hunk 4]
```diff
@@ -4,74 +4,113 @@
 def introduce_parser_arguments(
     parser: argparse.ArgumentParser,
 ) -> argparse.ArgumentParser:
-    # core hyperparameters, fixed for all experiments; needs a good reason to change
+    # CORE model hyperparameters, almost always specified in baseline configs
+    parser.add_argument("--model_type", type=str, choices=["gpt", "bert"])
+    parser.add_argument("--ff_mode", type=str, default="vanilla")
```
Comment (crewtool) [inline research/conditional/utils/argparse.py:9]: IMO we should go either no defaults or all defaults.

File: research/conditional/utils/argparse.py [hunk 5]
```diff
@@ -4,74 +4,113 @@
...
+    parser.add_argument("--n_blocks", type=int, required=True)
+    parser.add_argument("--dmodel", type=int, required=True)
+    parser.add_argument("--dff", type=int, required=True)
+    parser.add_argument("--n_att_heads", type=int, required=True)
+    parser.add_argument("--dhead", type=int, default=None)
```
Comment (crewtool) [inline research/conditional/utils/argparse.py:14]: Is default=None sam as required=False

File: research/conditional/utils/argparse.py [hunk 6]
```diff
@@ -4,74 +4,113 @@
...
 
-    parser.add_argument("--use_clearml", action="store_true")
-    parser.add_argument("--use_neptune", action="store_true")
-    parser.add_argument("--path_to_config", type=str, default=None)
-    parser.add_argument("--batch_size", type=int, default=600)
+    # other model hyperparameters
+    parser.add_argument("--activation_type", type=str, default="relu")
+    parser.add_argument("--residual_mode", type=str, default="pre_norm")
```
Comment (crewtool) [inline research/conditional/utils/argparse.py:18]: same

File: research/conditional/utils/argparse.py [hunk 7]
```diff
@@ -4,74 +4,113 @@
...
+    parser.add_argument("--every_other_layer", action="store_true")
+    parser.add_argument("--standard_ff_first", action="store_true")
+    parser.add_argument("--no_ff", action="store_true")
```
Comment (crewtool) [inline research/conditional/utils/argparse.py:21]: If those are always specified in baseline configs, then

File: research/conditional/utils/argparse.py [hunk 8]
```diff
@@ -4,74 +4,113 @@
...
+    # other training hyperparameters
+
+    parser.add_argument("--deterministic_experiment", action="store_true")
+    parser.add_argument("--adam_beta1", type=float, default=0.9)
+    parser.add_argument("--adam_beta2", type=float, default=0.999)
+    parser.add_argument("--grad_clip", type=float, default=None)
+    parser.add_argument("--weight_decay", type=float, default=0.0)
+    parser.add_argument("--lr_decay", type=float, default=None)
+    parser.add_argument("--lr_warmup_steps", type=int, default=0)
+    parser.add_argument("--lr_decay_interval", type=int, default=0)
```
Comment (crewtool) [inline research/conditional/utils/argparse.py:37]: as eralier one place for default

**Review by crewtool — COMMENTED** [review]
I believe we should specify baselines and no defaults

File: research/conditional/utils/argparse.py [hunk 9]
```diff
@@ -4,74 +4,113 @@
...
+    parser.add_argument("--dff", type=int, required=True)
+    parser.add_argument("--n_att_heads", type=int, required=True)
+    parser.add_argument("--dhead", type=int, default=None)
 
-    parser.add_argument("--use_clearml", action="store_true")
-    parser.add_argument("--use_neptune", action="store_true")
-    parser.add_argument("--path_to_config", type=str, default=None)
-    parser.add_argument("--batch_size", type=int, default=600)
+    # other model hyperparameters
+    parser.add_argument("--activation_type", type=str, default="relu")
```
Comment (crewtool) [inline research/conditional/utils/argparse.py:17]: usuwamy stąd

Comment (crewtool) [reply research/conditional/utils/argparse.py:17]: defautl

File: research/conditional/train/configs/test/test_baseline.yaml (hunk 3 above)
Comment (sebastianjaszczur) [reply research/conditional/train/configs/test/test_baseline.yaml:34]: Ale to jest tylko na unittesty, tak? To wtedy nie ma znaczenia, jak mniej więcej się zgadza. Tylko najwyżej dla pewności bym trzymał dmodel/dff jako śmiesznie mały (16/64), żeby nikt tego przypadkiem nie użył serio.

**Comment by sebastianjaszczur** [comment]
LGTM, jak komentarze Michała będą uwzględnione. Nie mogę się doczekać merge.

File: research/conditional/utils/argparse.py (hunk 9 above)
Comment (Simontwice) [reply research/conditional/utils/argparse.py:19]: defaulty zostaja na razie

File: research/conditional/train/configs/test/test_baseline.yaml (hunk 3 above)
Comment (Simontwice) [reply research/conditional/train/configs/test/test_baseline.yaml:33]: to unit test

File: research/conditional/utils/argparse.py (hunk 5 above)
Comment (Simontwice) [reply research/conditional/utils/argparse.py:16]: no; btw I have removed all required=False

File: research/conditional/utils/argparse.py (hunk 4 above)
Comment (Simontwice) [reply research/conditional/utils/argparse.py]: not for now

Comment (Simontwice) [reply research/conditional/utils/argparse.py]: trello task added btw

**Review by crewtool — APPROVED** [review]

**Pull request merged**